File: android/src/main/java/com/visioncameramlkit/infrastructure/mlkit/factories/BarcodeScanningServiceFactory.kt
package com.visioncameramlkit.infrastructure.mlkit.factories

import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.visioncameramlkit.domain.models.BarcodeFormatOption
import com.visioncameramlkit.domain.models.BarcodeScanningOptions
import com.visioncameramlkit.infrastructure.mlkit.MLKitBarcodeScanningService

object BarcodeScanningServiceFactory {
  fun create(options: BarcodeScanningOptions = BarcodeScanningOptions()): MLKitBarcodeScanningService {
    val scannerOptions = buildScannerOptions(options)
    val barcodeScanner = BarcodeScanning.getClient(scannerOptions)
    return MLKitBarcodeScanningService(barcodeScanner)
  }

  private fun buildScannerOptions(options: BarcodeScanningOptions): BarcodeScannerOptions {
    val formats = resolveFormats(options.formats)
    val builder =
      BarcodeScannerOptions
        .Builder()
        .setBarcodeFormats(formats.first(), *formats.drop(1).toIntArray())

    if (options.enableAllPotentialBarcodes) {
      builder.enableAllPotentialBarcodes()
    }

    return builder.build()
  }

  private fun resolveFormats(formats: List<BarcodeFormatOption>): List<Int> {
    if (formats.isEmpty() || formats.contains(BarcodeFormatOption.ALL)) {
      return listOf(Barcode.FORMAT_ALL_FORMATS)
    }

    val resolved = formats.mapNotNull { toMlKitFormat(it) }
    return if (resolved.isEmpty()) {
      listOf(Barcode.FORMAT_ALL_FORMATS)
    } else {
      resolved
    }
  }

  private fun toMlKitFormat(format: BarcodeFormatOption): Int? =
    when (format) {
      BarcodeFormatOption.CODE_128 -> Barcode.FORMAT_CODE_128
      BarcodeFormatOption.CODE_39 -> Barcode.FORMAT_CODE_39
      BarcodeFormatOption.CODE_93 -> Barcode.FORMAT_CODE_93
      BarcodeFormatOption.CODABAR -> Barcode.FORMAT_CODABAR
      BarcodeFormatOption.DATA_MATRIX -> Barcode.FORMAT_DATA_MATRIX
      BarcodeFormatOption.EAN_13 -> Barcode.FORMAT_EAN_13
      BarcodeFormatOption.EAN_8 -> Barcode.FORMAT_EAN_8
      BarcodeFormatOption.ITF -> Barcode.FORMAT_ITF
      BarcodeFormatOption.QR_CODE -> Barcode.FORMAT_QR_CODE
      BarcodeFormatOption.UPC_A -> Barcode.FORMAT_UPC_A
      BarcodeFormatOption.UPC_E -> Barcode.FORMAT_UPC_E
      BarcodeFormatOption.PDF417 -> Barcode.FORMAT_PDF417
      BarcodeFormatOption.AZTEC -> Barcode.FORMAT_AZTEC
      BarcodeFormatOption.ALL -> Barcode.FORMAT_ALL_FORMATS
      BarcodeFormatOption.UNKNOWN -> null
    }
}

File: ios/Infrastructure/MLKit/Factories/BarcodeScanningServiceFactory.swift
import Foundation

#if MLKIT_BARCODE_SCANNING
  import MLKitBarcodeScanning

  class BarcodeScanningServiceFactory {

    static func createBarcodeScanner(options: BarcodeScanningOptions)
      -> BarcodeScanner
    {
      let scannerOptions: MLKitBarcodeScanning.BarcodeScannerOptions
      if options.formats.isEmpty || options.formats.contains(.formatAllFormats) {
        scannerOptions = MLKitBarcodeScanning.BarcodeScannerOptions()
      } else {
        scannerOptions = MLKitBarcodeScanning.BarcodeScannerOptions(
          formats: toMLKitFormats(options.formats)
        )
      }
      return BarcodeScanner.barcodeScanner(options: scannerOptions)
    }

    static func create(options: BarcodeScanningOptions = BarcodeScanningOptions())
      -> MLKitBarcodeScanningService
    {
      let scanner = createBarcodeScanner(options: options)
      return MLKitBarcodeScanningService(barcodeScanner: scanner)
    }

    private static func toMLKitFormats(_ formats: [BarcodeFormatOption])
      -> MLKitBarcodeScanning.BarcodeFormat
    {
      var rawValue = 0
      for format in formats {
        switch format {
        case .formatUnknown, .formatAllFormats:
          continue
        case .formatCode128:
          rawValue |= 1
        case .formatCode39:
          rawValue |= 2
        case .formatCode93:
          rawValue |= 4
        case .formatCodabar:
          rawValue |= 8
        case .formatDataMatrix:
          rawValue |= 16
        case .formatEan13:
          rawValue |= 32
        case .formatEan8:
          rawValue |= 64
        case .formatItf:
          rawValue |= 128
        case .formatQrCode:
          rawValue |= 256
        case .formatUpcA:
          rawValue |= 512
        case .formatUpcE:
          rawValue |= 1024
        case .formatPdf417:
          rawValue |= 2048
        case .formatAztec:
          rawValue |= 4096
        }
      }

      return MLKitBarcodeScanning.BarcodeFormat(rawValue: rawValue)
    }
  }
#endif  // MLKIT_BARCODE_SCANNING
